### PR TITLE
WT-3850 Implement WT_SESSION::prepare_transaction

### PR DESCRIPTION
**Pull request by bvpvamsikrishna**
Raising this pull request for intermediate reviews, not ready for merge.
The changes are to implement and update the below functions to support prepared transactions. 
prepare_transaction
commit_transaction
rollback_transaction

Read / Update operation support of prepared transactions in not part of this ticket.

### CODE REVIEW

File: src/session/session_api.c [hunk 1]
```diff
@@ -1536,12 +1536,24 @@ __session_prepare_transaction(WT_SESSION *wt_session, const char *config)
...
 	WT_SESSION_IMPL *session;
+	WT_TXN *txn;
 
 	session = (WT_SESSION_IMPL *)wt_session;
 	SESSION_API_CALL(session, prepare_transaction, config, cfg, false);
 
 	WT_ERR(__wt_txn_context_check(session, true));
 
+	/*
+	 * Failed transaction cannot be prepared, as it cannot guarantee
```
Comment (agorrod) [inline src/session/session_api.c]: Failed -> A failed

File: src/txn/txn.c [hunk 2]
```diff
@@ -750,7 +750,9 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
 		WT_ERR_MSG(session, EINVAL, "commit_timestamp requires a "
 		    "version of WiredTiger built with timestamp support");
 #endif
-	}
+	} else if (F_ISSET(txn, WT_TXN_PREPARE))
+		WT_ERR_MSG(session, EINVAL, "commit_timestamp is required "
+		    "for a prepared transaction");
```
Comment (agorrod) [inline src/txn/txn.c]: Does the string fit on a single line?

File: src/txn/txn.c [hunk 3]
```diff
@@ -806,7 +808,7 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
 	 * positioned cursors so they don't point to updates that could be
 	 * freed once we don't have a snapshot.
```
Comment (agorrod) [inline src/txn/txn.c]: It would be good to explain why we don't do this for prepared transactions.

File: src/session/session_api.c [hunk 4]
```diff
@@ -1516,10 +1516,10 @@ __session_commit_transaction(WT_SESSION *wt_session, const char *config)
 		    txn->rollback_reason == NULL ? "" : ": ",
 		    txn->rollback_reason == NULL ? "" : txn->rollback_reason);
 
-	__wt_txn_prepare_clear(session);
 	if (ret == 0)
 		ret = __wt_txn_commit(session, cfg);
 	else {
+		__wt_txn_prepare_clear(session);
 		WT_TRET(__wt_session_reset_cursors(session, false));
```
Comment (agorrod) [inline src/session/session_api.c]:
```
if (!F_ISSET(WT_TXN_PREPARE))
    __wt_session_reset_cursors(session, false);
else
    WT_ASSERT(session, __ensure_all_cursors_reset(session));
```

File: src/txn/txn.c [hunk 5]
```diff
@@ -864,8 +866,30 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
...
+				    WT_UPDATE_PREPARE)) {
+					FLD_SET(op->u.upd->state,
+					    WT_UPDATE_LOCKED);
+
+					__wt_timestamp_set(
+					    &op->u.upd->timestamp,
+					    &txn->commit_timestamp);
+
+					FLD_SET(op->u.upd->state,
+					    WT_UPDATE_NONE);
```
Comment (agorrod) [inline src/txn/txn.c]: Does the unlock need to happen after all updates have had their timestamps reset?

File: src/txn/txn.c [hunk 6]
```diff
@@ -988,22 +1013,127 @@ int
...
+	/*
+	 * Look for a prepare timestamp.
+	 */
+	WT_ERR(
+	    __wt_config_gets_def(session, cfg, "prepare_timestamp", 0, &cval));
+	if (cval.len != 0) {
+		WT_ERR(__wt_txn_parse_timestamp(session,
+		    "prepare", &ts, &cval));
+
+		/* Validate prepare timestamp.  */
```
Comment (agorrod) [inline src/txn/txn.c]: `XXX` or `TODO:` are the strings I use for this sort of note to self.

File: src/txn/txn.c (hunk 5 above)
Comment (bvpvamsikrishna) [reply src/txn/txn.c]: Agreed in discussion that this unlock is ok, but need to make this logic as a function call.

**Comment by michaelcahill** [comment]
@bvpvamsikrishna, please note that I changed this pull request to be against `develop`, since the original base branch has now been merged.  I also merged `develop` into this branch to resolve conflicts.  Please let me know if I broke anything: I was just trying to clean up pull requests and branches in GitHub.

File: src/session/session_api.c [hunk 7]
```diff
@@ -1674,14 +1673,25 @@ __session_prepare_transaction(WT_SESSION *wt_session, const char *config)
 {
```
Comment (bvpvamsikrishna) [inline src/session/session_api.c]: prepare_transaction should be a "notsup" error in case of readonly connection.

**Comment by agorrod** [comment]
@bvpvamsikrishna Thanks  - this lgtm now, please go ahead and merge if it passes testing.

**Pull request merged**